packages/grpc-js/src/server-call.ts

```diff
@@ -146,3 +253,218 @@ export type Handler<RequestType, ResponseType> = {
+        }
+      });
+    });
+  }
+
+  serializeMessage(value: ResponseType) {
+    const handler = this.handler as Handler<RequestType, ResponseType>;
+    const messageBuffer = handler.serialize(value);
+
+    // TODO(cjihrig): Call compression aware serializeMessage().
+    const byteLength = messageBuffer.byteLength;
+    const output = Buffer.allocUnsafe(byteLength + 5);
+    output.writeUInt8(0, 0);
+    output.writeUInt32BE(byteLength, 1);
+    messageBuffer.copy(output, 5);
+    return output;
+  }
+
+  async deserializeMessage(bytes: Buffer) {
+    const handler = this.handler as Handler<RequestType, ResponseType>;
+    // TODO(cjihrig): Call compression aware deserializeMessage().
+    const receivedMessage = bytes.slice(5);
+
+    return handler.deserialize(receivedMessage);
+  }
+
+  async sendUnaryMessage(
+      err: ServiceError|null, value: ResponseType|null, metadata?: Metadata,
+      flags?: number) {
+    if (err) {
+      if (metadata) {
+        err.metadata = metadata;
+      }
+
+      this.sendError(err);
+      return;
+    }
+
+    try {
+      const response = await this.serializeMessage(value as ResponseType);
+
+      this.write(response);
+      this.sendStatus(
+          {code: Status.OK, details: 'OK', metadata} as StatusObject);
+    } catch (err) {
+      err.code = Status.INTERNAL;
+      this.sendError(err);
+    }
+  }
+
+  sendStatus(statusObj: StatusObject) {
+    if (this.cancelled === true) {
+      return;
+    }
+
+    if (this.deadline !== null) {
+      clearTimeout(this.deadline);
+      this.deadline = null;
+    }
+
+    if (!this.wantTrailers) {
+      this.wantTrailers = true;
+      this.stream.once('wantTrailers', () => {
+        let trailersToSend = {
+          [GRPC_STATUS_HEADER]: statusObj.code,
+          [GRPC_MESSAGE_HEADER]: encodeURI(statusObj.details as string)
+        };
+        const metadata = statusObj.metadata || this.metadata;
+
+        if (metadata) {
+          trailersToSend =
+              Object.assign(trailersToSend, metadata.toHttp2Headers());
+        }
+
+        this.stream.sendTrailers(trailersToSend);
+      });
+      this.sendMetadata();
+      this.stream.end();
+    }
+  }
+
+  sendError(error: ServiceError) {
+    const status: StatusObject = {
+      code: Status.UNKNOWN,
+      details: error.hasOwnProperty('message') ? error.message :
+                                                 'Unknown Error',
+      metadata: error.hasOwnProperty('metadata') ? error.metadata :
+                                                   this.metadata as Metadata
+    };
+
+    if (error.hasOwnProperty('code') && Number.isInteger(error.code)) {
+      status.code = error.code;
+
+      if (error.hasOwnProperty('details')) {
+        status.details = error.details;
+      }
+    }
+
+    this.sendStatus(status);
+  }
+
+  write(chunk: Buffer) {
+    if (this.cancelled === true) {
+      return;
+    }
+
+    this.sendMetadata();
+    return this.stream.write(chunk);
```

Comment:
This propagates the return value of `this.stream.write`, but it looks like it doesn't propagate the `'drain'` event, which the next layer up listens for.